src/shaders/pbr.frag

```diff
@@ -0,0 +1,292 @@
+      max(4.0 * max(dot(normal, light), 0.0) * max(dot(normal, view), 0.0),
+          Epsilon);
+  vec3 specular = Fresnel *
+                  specularGeometricAttenuation(normal, light, view, roughness) *
+                  normalDistribution(normal, half, roughness) / temp;
+
+  return (diffuse + specular) * lightRadiance * max(dot(normal, light), 0.0);
+}
+
+void main() {
+  vec3 emissiveColor = Material.emissiveColor;
+#if defined(EMISSIVE_TEXTURE)
+  emissiveColor *= texture(emissiveTexture, texCoord).rgb;
+#endif
+  fragmentColor = vec4(emissiveColor, 0.0);
+
+#if (LIGHT_COUNT > 0)
+  vec4 baseColor = Material.baseColor;
+#if defined(BASECOLOR_TEXTURE)
+  baseColor *= texture(baseColorTexture, texCoord);
+#endif
+
+  float roughness = Material.roughness;
+#if defined(NONE_ROUGHNESS_METALLIC_TEXTURE) || defined(OCCLUSION_ROUGHNESS_METALLIC_TEXTURE)
+  roughness *= texture(packedTexture, texCoord).g;
+#elif defined(ROUGHNESS_TEXTURE)
+  roughness *= texture(roughnessTexture, texCoord).r;
+#endif
+
+  float metallic = Material.metallic;
+#if defined(NONE_ROUGHNESS_METALLIC_TEXTURE) || defined(OCCLUSION_ROUGHNESS_METALLIC_TEXTURE)
+  metallic *= texture(packedTexture, texCoord).b;
+#elif defined(METALLIC_TEXTURE)
+  metallic *= texture(metallicTexture, texCoord).r;
+#endif
+
+#if defined(NORMAL_TEXTURE)
+  // normal is now in the camera space
+  vec3 n = getNormalFromNormalMap();
+#else
+  vec3 n = normal;
+#endif
+
+  // view dirction: a vector from current position to camera
+  // in camera space, camera is at the origin
+  vec3 view = normalize(-position);
+
+  vec3 finalColor = vec3(0.0);
+  highp float finalAlpha = 0.0;
+  // compute contribution of each light using the microfacet model
+  // the following part of the code is inspired by the Phong.frag in Magnum
+  // library (https://magnum.graphics/)
+  for (int iLight = 0; iLight < LIGHT_COUNT; ++iLight) {
+    // Attenuation. Directional lights have the .w component set to 0, use
+    // that to make the distance zero -- which will then ensure the
+    // attenuation is always 1.0
+    highp float dist =
+        length(LightDirections[iLight].xyz) * LightDirections[iLight].w;
+    // If range is 0 for whatever reason, clamp it to a small value to
+    // avoid a NaN when dist is 0 as well (which is the case for
+    // directional lights).
+    highp float attenuation =
+        clamp(1.0 - pow(dist / max(LightRanges[iLight], 0.0001), 4.0), 0.0,
+              1.0) /
+        (1.0 + dist);
+    attenuation = attenuation * attenuation;
+
+    // radiance
+    vec3 lightRadiance = LightColors[iLight] * attenuation;
+
+    // light source direction: a vector from current position to the light
+    vec3 light = normalize(LightDirections[iLight].xyz -
+                           position * LightDirections[iLight].w);
+
+    finalColor += microfacetModel(baseColor.rgb, metallic, roughness, n, light,
+                                  view, lightRadiance);
+    finalAlpha += baseColor.w;
```

Comment:
I don't understand this line.

Reply:
@eundersander: `.w` should between 0.0 and 1.0. Correct? :)

Reply:
Eric means it's an useless operation, you could as well do just 

```glsl
fragmentColor += vec4(finalColor, baseColor.a);
```

below and don't calculate any `finalAlpha`.

Reply:
Oh, OK, I see.
The add should happen outside the for loop since baseColor.a never changes.
fragmentColor += vec4(finalColor, baseColor.a);